File: Exercises/rb101-rb109_small_problems/medium_2/bubble_sort.rb
=begin

# Problem
- input: array of integers or strings
- output: array of integers or strings, sorted
- rules:
  - sort is in-place
  - input array contains at least two elements
  - array elements can be integers or strings, but not both

# Algorithm
Approach:
  - Iterate over the input array using `map`
  - Compare the current element of the array with the next element of the array
  - Swap if the first element is greater than the second; otherwise, leave in place.

1. Write a method that takes an array `array` as an argument.
2. Call `map!` and `with_index` on `array` with parameters `el` and `index`.
  2.1. el[index], el[index + 1] = el[index + 1], el[index] if el[index] > el[index + 1]
  2.2 break if el[index + 1] == array.index(array.last)

=end

def bubble_sort!(array)
  loop do
    greater_than = false
    (array.size - 1).times do |index|
      if array[index] > array[index + 1]
        array[index], array[index + 1] = array[index + 1], array[index]
        greater_than = true
      end
    end
    return array if greater_than == false
  end
end

array = [5, 3]
bubble_sort!(array)
p array == [3, 5]

array = [6, 2, 7, 1, 4]
bubble_sort!(array)
p array == [1, 2, 4, 6, 7]

array = %w(Sue Pete Alice Tyler Rachel Kim Bonnie)
bubble_sort!(array)
p array == %w(Alice Bonnie Kim Pete Rachel Sue Tyler)

File: ls215/exercises/small_problems/medium_2/bubble_sort.js
/* eslint-disable */

/*
# Problem
-input: array
  - array can contain numbers or strings
- output: array
- rules:
  - take an array of elements as input
  - sort array elements using a bubble sort algorithm
  - bubble sort algorithm:
    - iterate through the array from left to right
    - compare element at index and index + 1
    - if element at index + 1 is less than the element at index, swap the elements
    - continue iterating through the array, comparing each element at index with the element at index + 1
    - when the entire array can be iterated without swapping any elements, the array is sorted.

- questions:
  - can the array contain elements of different types?

# Data Structures
- input: array
- working with input: array
- output: array

# Algorithm
- take an array as input
- create a swapCount variable to track the number of swaps
- create an outer loop that iterates until no more elements are swapped 
  - create an inner loop that iterates over the elements of the input array
    - if the element at index is greater than the element at index + 1:
      - swap the elements' positions
      - increment swapCount variable by 1
  - if swapCount is > 0, continue to next iteration
  - if swapCount = 0, break out of the loop
  - return the sorted array

*/

/* eslint-enable */

function bubbleSort(array) {
  let swapCount;

  do {
    swapCount = 0;
    for (let index = 0; index < array.length; index += 1) {
      if (array[index] > array[index + 1]) {
        [ array[index], array[index + 1] ] = [array[index + 1], array[index]];
        swapCount += 1;
      }
    }
  } while (swapCount > 0);

  return array;
}

const array1 = [5, 3];
bubbleSort(array1);
console.log(array1);    // [3, 5]

const array2 = [6, 2, 7, 1, 4];
bubbleSort(array2);
console.log(array2);    // [1, 2, 4, 6, 7]

const array3 = ['Sue', 'Pete', 'Alice', 'Tyler', 'Rachel', 'Kim', 'Bonnie'];
bubbleSort(array3);
console.log(array3);    // ["Alice", "Bonnie", "Kim", "Pete", "Rachel", "Sue", "Tyler"]
